### PR TITLE
refactor: serial normalisation in SLE

### PR DESCRIPTION
**Pull request by noahjacob**
## Issue
- SLEs and other docs store multiple serial numbers in one field which break 1NF.

## Changes
- The serial no. field should have a single serial in a SLE.
- Creating separate SLEs for each serial. 

serial_norm

## To Do
- [x] Handle SLE creation for Buying.
- [x] Handle SLE creation for Selling.
- [x] Handle SLE creation for Stock Reco.
- [x] Handle SLE creating when cancelling Stock Reco.
- [x] Fix Stock Reco Report - Balance Serial is incorrect [#30041](https://github.com/frappe/erpnext/pull/30041)
- [x] Fix Edge cases and tests related to stock entry and backdated stock recos.
- [x] Fix tests that consider old sle serial format when asserting values.
- [x] Tests for new sles created with single serials - Buying
- [x] Tests for new sles created with single serials - Selling
- [x] Tests for new sles created with single serials - Stock Entry
- [x] Tests for new sles created with single serials - Stock Reco
- [x] Cleanup serial items table update functions.
- [ ]  Fix incorrect gle amounts posted

### CODE REVIEW

File: erpnext/controllers/buying_controller.py [hunk 1]
```diff
@@ -704,6 +706,72 @@ def validate_items(self):
...
+								break
+				if item.rejected_serial_no:
+					for serial in item.rejected_serial_no.split('\n'):
+						for serial_item in self.serial_items:
+							if serial_item.item_name == item.item_code and not serial_item.serial_no\
+								and serial_item.type == 'Rejected':
+								serial_item.serial_no = serial.upper()
+								frappe.db.set_value(serial_item.doctype, serial_item.name, 'serial_no', serial.upper())
+								break
+			return
```
Comment (marination) [inline erpnext/controllers/buying_controller.py]: - This is way too much indentation. Im going to leave it at that. (throughout this PR)
- The lone return indicates that this could maybe be a different function.

File: erpnext/accounts/doctype/purchase_invoice/purchase_invoice.json [hunk 2]
```diff
@@ -1417,13 +1419,26 @@
    "label": "Advance Tax",
    "options": "Advance Tax",
    "read_only": 1
+  },
+  {
+   "fieldname": "serial_items",
+   "fieldtype": "Table",
+   "label": "Serial Items",
+   "options": "Serial Number Item",
```
Comment (ankush) [inline erpnext/accounts/doctype/purchase_invoice/purchase_invoice.json]: can't find use of this table in any business logic? is this only for display?

File: erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py [hunk 3]
```diff
@@ -434,6 +434,8 @@ def on_submit(self):
 			self.set_consumed_qty_in_po()
 			from erpnext.stock.doctype.serial_no.serial_no import update_serial_nos_after_submit
 			update_serial_nos_after_submit(self, "items")
+			self.reload()
```
Comment (ankush) [inline erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py]: reload during submission is dangerous, all modifications to objects are lost. Why is this required?

File: erpnext/controllers/buying_controller.py [hunk 4]
```diff
@@ -704,6 +706,72 @@ def validate_items(self):
...
+
+		self.serial_items = {}
```
Comment (ankush) [inline erpnext/controllers/buying_controller.py]: child tables are list type not dict.

Comment (ankush) [reply erpnext/controllers/buying_controller.py]: also very long function.

File: erpnext/controllers/selling_controller.py [hunk 5]
```diff
@@ -585,6 +587,37 @@ def validate_items(self):
 		from erpnext.controllers.buying_controller import validate_item_type
 		validate_item_type(self, "is_sales_item", "sales")
 
+	def update_serial_items_table(self, update=False):
+		# To update auto-created serials.
+		if self.serial_items and update:
+			for item in self.items:
+				if item.serial_no:
+					for serial in item.serial_no.split('\n'):
```
Comment (ankush) [inline erpnext/controllers/selling_controller.py]:
```suggestion
					for serial in item.serial_no.split('\n'):
```

use utility function for getting a serial no from the string. Every manually handling of this has resulted in bugs in past.

File: erpnext/controllers/selling_controller.py [hunk 6]
```diff
@@ -585,6 +587,37 @@ def validate_items(self):
...
+						for serial_item in self.serial_items:
+							if serial_item.item_name == item.item_code and not serial_item.serial_no\
+								and serial_item.type == 'Accepted':
+								serial_item.serial_no = serial.upper()
+								frappe.db.set_value(serial_item.doctype, serial_item.name, 'serial_no', serial.upper())
+								break
```
Comment (ankush) [inline erpnext/controllers/selling_controller.py]: long, over-indented, duplicate code 😅

File: erpnext/controllers/stock_controller.py [hunk 7]
```diff
@@ -515,6 +515,65 @@ def repost_future_sle_and_gle(self):
...
+		if self.get(table_name) and update:
+			for item in self.items:
+				if item.serial_no:
+					for serial in item.serial_no.split("\n"):
+						for serial_item in self.get(table_name):
+							if serial_item.item_name == item.item_code and not serial_item.serial_no\
+								and serial_item.type == "Accepted":
+								serial_item.serial_no = serial.upper()
+								frappe.db.set_value(serial_item.doctype, serial_item.name, "serial_no", serial.upper())
+								break
```
Comment (ankush) [inline erpnext/controllers/stock_controller.py]: Third time very similar code appeared in same PR 😅

File: erpnext/controllers/stock_controller.py [hunk 8]
```diff
@@ -515,6 +515,65 @@ def repost_future_sle_and_gle(self):
...
+			self.new_serials = {}
+		else:
+			self.serial_items = {}
+
+		for item in self.items:
+			has_serial = frappe.db.get_value("Item", item.item_code, "has_serial_no")
+
+			# Updates current_serials table in stock reco.
+			if has_serial and item.get("current_serial_no"):
+				current_serials = item.current_serial_no.split("\n")
```
Comment (ankush) [inline erpnext/controllers/stock_controller.py]: manual splitting of serial nos.

File: erpnext/stock/doctype/stock_ledger_entry/stock_ledger_entry.json [hunk 9]
```diff
@@ -309,6 +311,22 @@
    "label": "Recalculate Incoming/Outgoing Rate",
    "no_copy": 1,
    "read_only": 1
+  },
+  {
+   "default": "0",
+   "fieldname": "skip_update_serial_no",
```
Comment (ankush) [inline erpnext/stock/doctype/stock_ledger_entry/stock_ledger_entry.json:317]: This doesn't need to go in database, can just be a flag on SLE object during insertion.

File: erpnext/stock/doctype/stock_reconciliation/stock_reconciliation.py [hunk 10]
```diff
@@ -236,14 +244,42 @@ def update_stock_ledger(self):
...
+						continue
+					if flt(sle.get('actual_qty')) > 1:
+						# If it needs to auto create serials
+						serials = get_serial_nos(sle.serial_no)
+						item_det = get_item_details(sle.get('item_code'))
+						validate_serial_no(sle, item_det)
+						update_serial_nos(sle, item_det)
+
+						qty_after_transaction = 0
+						for serial in range(len(serials)):
```
Comment (ankush) [inline erpnext/stock/doctype/stock_reconciliation/stock_reconciliation.py]: why not just iterate directly? it's a list of strings.

File: erpnext/stock/stock_ledger.py [hunk 11]
```diff
@@ -65,22 +69,67 @@ def make_sl_entries(sl_entries, allow_negative_stock=False, via_landed_cost_vouc
...
+								if (item.s_warehouse or item.t_warehouse) == sle.warehouse:
+									frappe.db.set_value(item.doctype, item.name, 'serial_no', sle.serial_no)
+									break
+							if item.warehouse == sle.warehouse:
+								frappe.db.set_value(item.doctype, item.name, 'serial_no', sle.serial_no)
+							elif item.rejected_warehouse == sle.warehouse:
+								frappe.db.set_value(item.doctype, item.name, 'rejected_serial_no', sle.serial_no)
+							break
+					serials = sle.serial_no.strip().split('\n')
+					for serial in range(abs(cint(sle['actual_qty']))):
```
Comment (ankush) [inline erpnext/stock/stock_ledger.py]: iterating over qty range and assigning a serial number by popping, why not just iterate over serial nos? 

Also, manual splitting of the serial string is problematic.

File: erpnext/stock/doctype/serial_number_item/serial_number_item.json [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+   "label": "Type"
+  }
+ ],
+ "index_web_pages_for_search": 1,
+ "istable": 1,
+ "links": [],
+ "modified": "2022-02-16 14:39:09.711465",
+ "modified_by": "Administrator",
+ "module": "Stock",
+ "name": "Serial Number Item",
```
Comment (marination) [inline erpnext/stock/doctype/serial_number_item/serial_number_item.json]: Why not "Serial No Item", naming is inconsistent with Serial No DocType

File: erpnext/controllers/buying_controller.py (hunk 1 above)
Comment (noahjacob) [reply erpnext/controllers/buying_controller.py]: Yea I'm trying to find a better way to do this.

File: erpnext/stock/stock_ledger.py [hunk 13]
```diff
@@ -65,22 +69,67 @@ def make_sl_entries(sl_entries, allow_negative_stock=False, via_landed_cost_vouc
...
+						new_sle = sle.copy()
+						new_sle['actual_qty'] = 1 if sle['actual_qty'] > 0 else -1
+						new_sle['serial_no'] = serials.pop()
+						new_sle['skip_update_serial_no'] = 1
+						new_sle['skip_serial_no_validation'] = 1
+						sle_docs.append(make_entry(new_sle, allow_negative_stock, via_landed_cost_voucher))
+						sr = frappe.get_doc('Serial No', new_sle['serial_no'])
+						sr.update_serial_no_reference(sr.serial_no, update=True)
```
Comment (ankush) [inline erpnext/stock/stock_ledger.py]: what is being updated here again? The voucher number will be the same regardless of single or multiple SLEs.

File: erpnext/stock/stock_ledger.py [hunk 14]
```diff
@@ -65,22 +69,67 @@ def make_sl_entries(sl_entries, allow_negative_stock=False, via_landed_cost_vouc
...
+					# preserve previous_qty_after_transaction for qty reposting
+					args.previous_qty_after_transaction = sle.get("previous_qty_after_transaction")
+
+				is_stock_item = frappe.get_cached_value('Item', args.get("item_code"), 'is_stock_item')
+				if is_stock_item:
+					bin_name = get_or_make_bin(args.get("item_code"), args.get("warehouse"))
+					repost_current_voucher(args, allow_negative_stock, via_landed_cost_voucher)
+					update_bin_qty(bin_name, args)
+				else:
+					frappe.msgprint(_("Item {0} ignored since it is not a stock item").format(args.get("item_code")))
```
Comment (ankush) [inline erpnext/stock/stock_ledger.py:94]: This function is now 100 lines long. All serial splitting logic can be easily moved to a separate function.

**Review by ankush — CHANGES_REQUESTED** [review]

**Comment by ankush** [comment]
Some more thoughts on the design:

- Do we intend to patch existing data? If not then we will continue to have this crazy query and all performance problems that come with it: https://github.com/frappe/erpnext/blob/7a7a7f283e6287326be4b8ef866c3eaf549a96cf/erpnext/stock/doctype/serial_no/serial_no.py#L166-L187

**Comment by ankush** [comment]
Accounting impact is wrong since only first SLE is getting considered for posting GLE:



Screenshot 2022-03-22 at 6 40 03 PM


https://github.com/frappe/erpnext/blob/7a7a7f283e6287326be4b8ef866c3eaf549a96cf/erpnext/stock/doctype/purchase_receipt/purchase_receipt.py#L289-L291


Essentially ALL `voucher_detail_no` related business logic has to be checked for this case.

File: erpnext/stock/doctype/purchase_receipt/purchase_receipt.py [hunk 15]
```diff
@@ -284,9 +284,14 @@ def make_item_gl_entries(self, gl_entries, warehouse_account=None):
...
-					stock_value_diff = frappe.db.get_value("Stock Ledger Entry",
+					stock_value_diff = frappe.db.get_values("Stock Ledger Entry",
 						{"voucher_type": "Purchase Receipt", "voucher_no": self.name,
-						"voucher_detail_no": d.name, "warehouse": d.warehouse, "is_cancelled": 0}, "stock_value_difference")
+						"voucher_detail_no": d.name, "warehouse": d.warehouse, "is_cancelled": 0}, "stock_value_difference", as_dict=True, pluck="stock_value_difference")
+					has_serial = frappe.db.get_value('Item', d.item_code, 'has_serial_no')
+					if has_serial:
+						stock_value_diff = sum(stock_value_diff)
+					else:
+						stock_value_diff = stock_value_diff[0]
```
Comment (ankush) [inline erpnext/stock/doctype/purchase_receipt/purchase_receipt.py]: It can always be the sum IMO. Make it future-proof. 

Alternatively `sum(stock_value_difference)` can be fetched directly.
